src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -358,6 +361,129 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
+                {
+                    var substituteGenericParameterConstraint = SubstituteGenericParameterConstraint(parameters, constraint);
+                    if (!ParameterCompatibleWithTypeConstraint(parameter, substituteGenericParameterConstraint))
+                    {
+                        return false;
+                    }
+                }
+
+                var specialConstraints = argumentDefinitionTypeInfo.GenericParameterAttributes;
+                if ((specialConstraints & GenericParameterAttributes.DefaultConstructorConstraint) == GenericParameterAttributes.DefaultConstructorConstraint)
+                {
+                    var hasPublicNoArgCtor = false;
+                    foreach (var c in parameterTypeInfo.DeclaredConstructors)
+                    {
+                        if (c.IsPublic && c.GetParameters().Length == 0)
+                        {
+                            hasPublicNoArgCtor = true;
+                            break;
+                        }
+                    }
+
+                    if (!parameterTypeInfo.IsValueType && !hasPublicNoArgCtor)
+                    {
+                        return false;
+                    }
+                }
+                if ((specialConstraints & GenericParameterAttributes.ReferenceTypeConstraint) == GenericParameterAttributes.ReferenceTypeConstraint)
+                {
+                    if (parameterTypeInfo.IsValueType)
+                    {
+                        return false;
+                    }
+                }
+                if ((specialConstraints & GenericParameterAttributes.NotNullableValueTypeConstraint) == GenericParameterAttributes.NotNullableValueTypeConstraint)
+                {
+                    if (!parameterTypeInfo.IsValueType || parameterTypeInfo.IsGenericType && IsGenericTypeDefinedBy(parameter, typeof(Nullable<>)))
+                    {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private static bool IsGenericTypeDefinedBy(Type type, Type openGeneric)
+        {
+            return !type.ContainsGenericParameters && type.IsGenericType && type.GetGenericTypeDefinition() == openGeneric;
+        }
+
+        private static Type SubstituteGenericParameterConstraint(Type[] parameters, Type constraint)
+        {
+            return constraint.IsGenericParameter ? parameters[constraint.GenericParameterPosition] : constraint;
+        }
+
+        private static bool ParameterCompatibleWithTypeConstraint(Type parameter, Type constraint)
+        {
+            if (constraint.IsAssignableFrom(parameter))
+            {
+                return true;
+            }
+
+            if (ParameterEqualsConstraint(parameter, constraint))
+            {
+                return true;
+            }
+
+            if (parameter.BaseType != null && ParameterEqualsConstraint(parameter.BaseType, constraint))
```

Comment:
It seems odd to me that you would check the parent type, but not the grandparent type, great-grandparent type, etc...

Reply:
@jbogard @halter73 
Chiming in here as I have some experience with this from LightInject. Actually this was one of the things that was really hard to get right as it would practically mimic the behaviour of the compiler. I looked at several other implementations  and if I remember correctly, they all do a `MakeGenericType` in the end to make sure. 

For what its worth, here is what I've got for LightInject 

https://github.com/seesharper/LightInject/blob/70cfc68718a57bf02959bf7a09fe59f0d372cae5/src/LightInject/LightInject.cs#L6808

Reply:
@halter73 that was my first go (see https://github.com/aspnet/DependencyInjection/pull/472/files#diff-599c17314f9888a15039a1f175fb6c35) - to only rely on MakeGenericType and not try to reproduce any of the generic type loading arguments. Honestly, I'd still rather go that route, as it's the most correct option. However, it makes exceptions part of the hot path and becomes a pretty big perf hit for constraint misses.

Reply:
Though I'm a still bit unsure of the overall performance hit, since hit's only the first resolution that the exception hits for constraint misses, and from then on out the resolution plan is cached.

Reply:
And to me the long-term solution is to remove all of this in favor of my API proposal https://github.com/dotnet/corefx/issues/33769 - `TryMakeGenericType`, since I've just tried to reproduce the easy cases with this approach. These checks, if successful, still get retried with a call to `MakeGenericType`.

Reply:
I agree with @halter73. Taking perf hit for an uncommon case to make sure things stay correct is acceptable.

Reply:
Alright, I'll push the simple, but comprehensive, but perf hit case. I prefer that as well - it will make things easy when (if?) `TryMakeGenericType` happens.